### PR TITLE
Avoid .valueOf to close #20594

### LINKED ISSUE #20594: Bug: unintended .valueOf() maybe?
While chasing https://github.com/tc39/proposal-temporal/issues/1224 where I mistakenly pasted a `ZonedDateTime` into JSX, I discovered that:

* `react-reconciler` doesn't like arbitrary objects (fair)
* calls `throwOnInvalidObjectType` on such objects, source code of which looks completely fine†
* but said source code gets compiled into `... "foo" + newChild + "bar" ...`‡
* at the same time, react explicitly chose not to `valueOf()` objects in https://github.com/facebook/react/issues/4769
* this specific type throws unconditionally on `.valueOf` (I'm ambivalent about that decision)

† https://github.com/facebook/react/blob/98313aaa7ea58d49eb30ecffc0f9eb6fc1ef467a/packages/react-reconciler/src/ReactChildFiber.new.js#L223-L235

‡ compiled to `throw Error( "Objects are not valid as a React child (found: " + (Object.prototype.toString.call(newChild) === '[object Object]' ? 'object with keys {' + Object.keys(newChild).join(', ') + '}' : newChild) + "). If you meant to render a collection of children, use an array instead." );`

So, react won't use `valueOf` to render an element, but will use it to report and error about an element.
Maybe the "compilation" or interpolation of `invariant` is wrong?
Or perhaps `%s` handling in `invariant` is wrong?
Maybe my assumptions are wrong?
Maybe that type is wrong?

React version: `17.0.1`

Steps To Reproduce

1. `return <div>{ new Temporal.Instant(0n).toZonedDateTimeISO("Europe/Berlin") }</div>;`

Link to code example: https://github.com/dimaqq/test-temporal

### PR DESCRIPTION
**Pull request by dimaqq**
close #20594

## Summary

`invariant` interpolates the arguments using `+`, which means `.valueOf` is called.
some types throw on `.valueOf` in order to prevent arithmetic operations (questionable, but still...)
this change 

## Test Plan

See #20594

### CODE REVIEW

File: packages/react-reconciler/src/ReactChildFiber.new.js [hunk 1]
```diff
@@ -222,14 +222,15 @@ function coerceRef(
 
 function throwOnInvalidObjectType(returnFiber: Fiber, newChild: Object) {
   if (returnFiber.type !== 'textarea') {
+    const child = Object.prototype.toString.call(newChild);
```
Comment (rickhanlonii) [inline packages/react-reconciler/src/ReactChildFiber.new.js:225]: How about `childString`?

**Comment by dimaqq** [comment]
sandbox build failed with `{ [Error: EACCES: permission denied, mkdir '/tmp/output/scheduler']` which I guess is some systemic issue. I'll try to rebase and hope for the best :)

**Review by rickhanlonii — APPROVED** [review]
LGTM

**Comment by acdlite** [comment]
So now if I return an instance of `MyCustomType`, and it `MyCustomType#toString` is defined, we'll no longer call it? Not a big deal, but seems unintentional.

Also maybe we shouldn't stringify the type in the production build at all. Only in dev.

**Comment by rickhanlonii** [comment]
@acdlite that was intentional on my part because calling the toString of the instances will give you some string representation of the value of the instance instead of a string of the object's name, so the error message doesn't make sense.

E.g.:

> Objects are not valid as a React child (found: 1611163658464)

vs:

> Objects are not valid as a React child (found: [object MyCustomType])

Is that fair?

**Pull request merged**